src/App.tsx

```diff
@@ -1,26 +1,259 @@
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
+  const [inputText, setInputText] = useState('');
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [tempTodos, setTempTodos] = useState<Todo[]>([]);
+  const [loadingError, setLoadingError] = useState('');
+  const [validationError, setValidationError] = useState('');
+  const [loadingTodos, setLoadingTodos] = useState<number[]>([]);
+
+  const inputRef = useRef<HTMLInputElement>(null);
+  const [filterSelected, setFilterSelected] = useState<Filter>(Filter.All);
+
+  const filterTodos = useMemo(() => {
+    switch (filterSelected) {
+      case Filter.Active:
+        return todos.filter(todo => !todo.completed);
+
+      case Filter.Completed:
+        return todos.filter(todo => todo.completed);
+
+      default:
+        return todos;
+    }
+  }, [todos, filterSelected]);
+
+  useEffect(() => {
+    getTodos()
+      .then(setTodos)
+      .catch(() => {
+        setLoadingError('Unable to load todos');
+        setTimeout(() => {
+          setLoadingError('');
+        }, 3000);
+      });
+  }, []);
+
+  useEffect(() => {
+    if (inputRef.current) {
+      inputRef.current.focus();
+    }
+  }, [todos, loadingTodos]);
+
+  const toggleTodo = (id: number) => {
+    setTodos(prevTodos =>
+      prevTodos.map(todo =>
+        todo.id === id ? { ...todo, completed: !todo.completed } : todo,
+      ),
+    );
+  };
+
+  const toggleTodoAll = () => {
+    const allCompleted = todos.every(todo => todo.completed);
+
+    setTodos(prevTodos =>
+      prevTodos.map(todo => ({ ...todo, completed: !allCompleted })),
+    );
+  };
+
+  const handleAddTodo = (e: React.FormEvent) => {
+    e.preventDefault();
+
+    if (!inputText.trim()) {
+      setValidationError('Title should not be empty');
+      setTimeout(() => {
+        setValidationError('');
+      }, 3000);
+      if (inputRef.current) {
+        inputRef.current.focus();
+      }
+
+      return;
+    }
+
+    const tempTodo: Todo = {
+      id: Date.now(),
+      title: inputText,
+      completed: false,
+      userId: 1129,
+    };
+
+    setTempTodos(prevTempTodos => [...prevTempTodos, tempTodo]);
+    setLoadingTodos(prev => {
+      const updated = [...prev, tempTodo.id];
+
+      return updated;
+    });
+
+    addTodo(inputText.trim())
+      .then(newTodo => {
+        setTodos(prevTodos => [...prevTodos, newTodo]);
+        setTempTodos(prevTempTodos =>
+          prevTempTodos.filter(todo => todo.id !== tempTodo.id),
+        );
+        setInputText('');
+        setValidationError('');
+        if (inputRef.current) {
+          inputRef.current.focus();
+        }
+      })
+      .catch(() => {
+        setLoadingError('Unable to add a todo');
+        setTempTodos(prevTempTodos =>
+          prevTempTodos.filter(todo => todo.id !== tempTodo.id),
+        );
+        setTimeout(() => {
+          setLoadingError('');
+        }, 3000);
+      })
+      .finally(() => {
+        setLoadingTodos(prev => {
```

Comment:
Since only one action can happen at a time, you can simplify it and just set empty array here